### DISCUSSION TITLE
debug/ developer tooling accomodations

### DISCUSSION THREAD

**Discussion by MACKAT05** [Snowflake-Labs/schemachange#396] [Ideas]
Developer tooling.
I wrote a simple PowerShell script to run the render command over all the files and redirect them into a subfolder that is listed in the .gitignore
Can we add Block quotes  to the outputs like so?:
![image](https://user-images.githubusercontent.com/12175865/219823310-acee68a2-948a-4c6d-98f7-a640aeebb2e1.png)

This would allow  Vs code users to execute and debug the rendered code interactively  as captured from the output with the new snowflake extension.

**Comment by MACKAT05** [comment]
Optionally this loop over command could be incorporated (and refined to match the deployment gathering Regular expressions)  into an expanded render command syntax... as per #156

**Comment by sfc-gh-tmathew** [comment]
Interesting @MACKAT05. Using the Render to verify the sql that will be executed appears to be useful application of the `render` functionality. It is as if we need a `--dry-run` option for the render command without incrementing the version of the SQL files to during the development lifecycle.

How do you think this pattern will fare for CI/CD based deployment workflows in Development environment? Are you envisioning the iterative development is done locally, say on a clone and then pushed for actual deployment in Development and higher environments?

**Comment by MACKAT05** [comment]
@sfc-gh-tmathew   i started working on a fork to deal with this and other issues...
* parallel execution for the jinja interpreter, full debug reporting
* new macro from_csv --> create a jinja python table to iterate over for use in higher level pipeline macro development.
* state detection of deployment --> eg what is deployed where among multiple targets
* strict option --> fail deployment if a skipped v script is detected or a regression of an R script is detected.
* rework snowflake deployment to use the underlying file iterator(https://github.com/snowflakedb/snowflake-connector-python/blob/ca87fd8412efebfbdeb34ff9fabe20b929da9f27/src/snowflake/connector/util_text.py#L35) so comments and empty statement can be ignored... refine activity detection so state normalization only occurs when needed.  which kicks open the door for more improvements.
 
Getting to the point of this query
while I was working on it I split cli.py into individual class files fore each level of functionality. I needed to do this to help isolate the levels of responsibility.  

 **I recall seeing somewhere that there was a technical reason for keeping it as one file... does that still need to happen?** 
or would a split of files like this ( with some naming oversight be a maintainable option)
![image](https://github.com/Snowflake-Labs/schemachange/assets/12175865/d4d767c3-6f92-4212-a9ce-c362a01a378a)

**Comment by sfc-gh-tmathew** [comment]
This appears to be a major refactor. Appears to be a 4.0 release candidate. Will get back to you @MACKAT05 once I have a chance to sync up with @jeremiahhansen.

**Comment by zanebclark** [comment]
@MACKAT05 , in  #275  I've split the cli.py file out into multiple files / directories to make it easier to write unit tests and such. I poked around your fork and couldn't find a branch with the folder structure above. Is that work in a local branch that you haven't committed? Do you think you could build your feature off of #275?

**Comment by MACKAT05** [comment]
yeah sorry about that it is still on my local... i'll try to get that visible and up later tonight.
edit added experimental branch and published the current state https://github.com/MACKAT05/schemachange/tree/experimental/schemachange

**Comment by sfc-gh-tmathew** [comment]
@MACKAT05 How can we breakdown the features you are trying to add to schemachange into one of the upcoming releases ?

**Comment by MACKAT05** [comment]
> [@MACKAT05](https://github.com/MACKAT05) How can we breakdown the features you are trying to add to schemachange into one of the upcoming releases ?

I had put some of them in PR's already earlier this year.  I ended up traveleing and getting sick and loosing all steam on them but i'm ready to pick them back up.   
It looks like from the traffic i have seen that there has been some work on improving the login/ auth integration.

Down stream of this in the program i would love to see ( and work on) exposing and improving the "planner" actions in how the program intends to deploy code.   As well as add some guard rails like regression detection and version code skip detection.
 
Expose the jijna renderer more directly so that its output could be used in a WYSWYG style interface along with funcitons that output the hash used in the logging.

Expand the logging functionality with dynamic definitions via settings in the schemachange.yml file. 

Improve the logging orchastration by running the logger on a seperate thread and connection. 

Ensure we don't end up with the same dbt faceplant that happens with multiple threads hitting up the snowflake auth and opening twenty chrome windows!

Expand the jijna functions to take in data files to enable table driven design. ( pr in place, i know i need to go over that)  

Expand the jijna functions to dynamically query the snowflake database for context design similar to dbtutils.get column values.

Expand the jijna functions to dynamically query databases (not inside of the snowflake connection) for anything to be used in jijna expansion.  (this could be used for automatic context updates.  like a table changed in the source system.

better context of files being deployed  (eg 2 tables created 1 view created.)
This would need better parsing of files (I have a PR that moves the parse and split out of the snowflake function into this programs control)

I think i saw some notion about some planning meetings... Please reach out and invite or direct me to the discussion board.

**Reply by sfc-gh-tmathew** [reply]
Your contributions are top quality and very useful. Glad you are over the sickness and ready to contribute again. Lets break it down into iterations as all your ideas will improve the usability of the tool..

**Comment by sfc-gh-tmathew** [comment]
@MACKAT05 love the ideas you have to build our roadmap and figure out how to release these features. Let me get back to you with some times to discuss. For now, I am moving this to the discussion forum to plan the breakdown and associate separate issues as you have listed above into specific milestones.
